### ISSUE TITLE
Call Timeout Implementation

### ISSUE THREAD

**Issue by ericchapman** [wamp-proto/wamp-proto#265]
@oberstet 
Hey Tobias,
I was trying to implement this in my Ruby library and had a few questions

```
A _timeout_ allows to *automatically* cancel a call after a specified time either at the _Callee_ or at the _Dealer_.
```

Since there is no sequence diagram, I read this as the client should perform the "CANCEL/INTERRUPT" flow when the timeout expires?  Is this correct?  If so, what should the "mode" of the cancel be?

Thanks
Eric

**Comment by oberstet** [comment]
@ericchapman Hi Eric, yeah, agreed: a sequence diagram would really be useful here. Also, to be honest, the text is lacking in general:(

The general idea is that there are two possible initiators of a timeout of a pending call: the original Caller and the Router. A router that has sent out an `INVOCATION` message to a Callee and does not a receive a `YIELD` within a time can fail the call and return a `CALL_ERROR` to the Callee. Or the Callee that has sent a `CALL` message to the Router and does not receive either a `CALL_RESULT` or `CALL_ERROR` within a time can fail the call and raise an exception in application code.

I think the former (the Router handling the timeout logic) is definitely useful. A client can simply set a value for call timeout, and the Router handles the annoying part. Apparantly, I originally thought there were reasons for the second variant (the Caller handling the timout logic), but right now, mmh;)

**Comment by ericchapman** [comment]
I agree with the router handling it.  I had it implemented as a timer going off on the client and the timer would send the cancel which would take it round trip and then eventually the client would get the error back.  Seems like overkill

**Comment by oberstet** [comment]
Yep, having the router take the burden of maintaining timers also is reusable across client libraries (does not need to be implemented in each and every client library).

Another thing that needs agreement: the error URI within the `CALL_ERROR` the Dealer sends back to the Client upon a timeout (on the `INVOCATION` message sent to the Callee and no `YIELD` being received in time).

Yet another thing: do we send an `INTERRUPT` to the Callee to inform it that there is no need to produce a result anymore (since the call was already timed out). Like with explicit call cancel: https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/advanced/ap_rpc_call_canceling.md

And finally: we can't keep the state for a canceled call around for infinite time, and hence, we'll loose the invocation request ID, and therefor, should a result come back from the Callee after the timeout, we can't distinguish that from a mere bogus invocation request ID.

https://github.com/wamp-proto/wamp-proto/blob/master/rfc/text/advanced/ap_rpc_call_timeout.md

**Comment by ericchapman** [comment]
I like the idea of sending the "INTERRUPT" because I think it takes care of allot of the "if"s since

 - the callee will be notified that the call has been cancelled (which "mode" should it be?  "skip" | "kill" | "killnowait")
 - an error will be sent to the client from the callee signalling that the call was cancelled

It seems to tie off any loose ends.  My vote goes to having the router trigger the "cancel" operation when the timeout is reached.

**Comment by gammazero** [comment]
I agree with @ericchapman about having the router trigger a `CANCEL` when the timeout is reached.  The "mode" should be "killnowait", which sends an `INTERRUPT` to the called so that the callee can potentially stop doing whatever long-running operation it is doing, and does not require the router to wait for a response (which may not be received if the callee has become non-responsive, which caused the timeout in the first place).
